Tracking prevention should only delete script-writable storage associated with a first-party

pooneh-nb opened the pull request:
#### 163a9bec9befb89cb3c9ae47b7fe2c93af5117a5
<pre>
Tracking prevention should only delete script-writable storage associated with a first-party
<a href="https://bugs.webkit.org/show_bug.cgi?id=262495">https://bugs.webkit.org/show_bug.cgi?id=262495</a>
<a href="https://rdar.apple.com/116144212">rdar://116144212</a>

Reviewed by NOBODY (OOPS!).

Tracking prevention&apos;s current behavior deletes all script-writable storage
without regard for the context where it was created. This means that all
storage associated with an origin will be deleted after 7 days of not visiting
the site as a first-party (in general). This deletion policy is not actually
correct for partitioned third-party storage because that data doesn&apos;t have any
connection with visiting the site as a first-party. Instead, we should only
delete script-writable storage associated with the first-party (where the
partition&apos;s top-level origin is the first-party).

* LayoutTests/http/tests/resourceLoadStatistics/resources/iframe-with-storage.html: Added.
* LayoutTests/http/tests/resourceLoadStatistics/resources/popup-with-storage-and-iframe.html: Added.
* LayoutTests/http/tests/resourceLoadStatistics/resources/popup-with-storage.html: Added.
* LayoutTests/http/tests/resourceLoadStatistics/resources/util.js:
(sortStringArray):
(addLinebreakToOutput):
(addOutput):
(checkCookies):
(addFrameEventListener):
(request.onupgradeneeded):
(initStorage):
(finish):
(receiveResponse):
(checkFrameStorage):
(checkIDBDataStoreExists):
(checkLocalStorageExists):
(async writeWebsiteDataAndContinue):
(async processWebsiteDataAndContinue):
(async checkWebsiteDataAndContinue):
(async finishTest):
* LayoutTests/http/tests/resourceLoadStatistics/resources/utilities.js: Added.
(setEnableFeature):
(async resetCookiesITP):
(sortStringArray):
(addLinebreakToOutput):
(addOutput):
(finish):
(receiveResponse):
(checkFrameStorage):
(addFrameEventListener):
(onMsg):
(checkPopupCookiesLikeCheckCookies):
(request.onupgradeneeded):
(createIDBDataStore):
(initStorage):
(checkIDBDataStoreExists):
(checkLocalStorageExists):
(once):
(async phaseA_writeWebsiteDataAndReturn.):
(async phaseA_writeWebsiteDataAndReturn):
(async processWebsiteDataAndContinue):
(async checkWebsiteDataAndContinue):
(async finishTest):
* LayoutTests/http/tests/resourceLoadStatistics/website-data-removal-for-site-navigated-to-with-filtered-link-decoration-after-long-deletion.html:
* LayoutTests/http/tests/resourceLoadStatistics/website-data-removal-for-site-with-user-interaction-and-cross-origin-without-user-interaction-expected.txt: Added.
* LayoutTests/http/tests/resourceLoadStatistics/website-data-removal-for-site-with-user-interaction-and-cross-origin-without-user-interaction.html: Added.
* LayoutTests/http/tests/resourceLoadStatistics/website-data-removal-for-site-with-user-interaction-expected.txt:
* LayoutTests/http/tests/resourceLoadStatistics/website-data-removal-for-site-with-user-interaction.html:
* LayoutTests/http/tests/resourceLoadStatistics/website-data-removal-for-site-without-interaction-and-retain-partitioned-and-unpartitioned-storage-for-site-with-interaction-expected.txt: Added.
* LayoutTests/http/tests/resourceLoadStatistics/website-data-removal-for-site-without-interaction-and-retain-partitioned-and-unpartitioned-storage-for-site-with-interaction.html: Added.
* LayoutTests/http/tests/resourceLoadStatistics/website-data-removal-for-site-without-user-interaction-but-with-user-interaction-on-cross-origin-domains-expected.txt: Added.
* LayoutTests/http/tests/resourceLoadStatistics/website-data-removal-for-site-without-user-interaction-but-with-user-interaction-on-cross-origin-domains.html: Added.
* LayoutTests/http/tests/resourceLoadStatistics/website-data-removal-for-site-without-user-interaction-expected.txt:
* LayoutTests/http/tests/resourceLoadStatistics/website-data-removal-for-site-without-user-interaction.html:
* LayoutTests/platform/glib/http/tests/resourceLoadStatistics/website-data-removal-for-site-without-interaction-and-retain-partitioned-and-unpartitioned-storage-for-site-with-interaction.html: Added.
* LayoutTests/platform/glib/http/tests/resourceLoadStatistics/website-data-removal-for-site-without-user-interaction-but-with-user-interaction-on-cross-origin-domains-exppected.txt: Added.
* LayoutTests/platform/glib/http/tests/resourceLoadStatistics/website-data-removal-for-site-without-user-interaction-expected.txt:
* Source/WebKit/NetworkProcess/storage/NetworkStorageManager.cpp:
(WebKit::NetworkStorageManager::deleteDataForRegistrableDomains):
</pre><!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/163a9bec9befb89cb3c9ae47b7fe2c93af5117a5

| Misc | iOS, visionOS, tvOS & watchOS  | macOS  | Linux |  Windows | Apple Internal |
| ----- | ---------------------- | ------- |  ----- |  --------- | ------ |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/120177 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/131/builds/39871 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/138/builds/30522 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/126527 "Built successfully") | [✅ 🛠 win](https://ews-build.webkit.org/#/builders/59/builds/72250 "Built successfully") | [❌ 🛠 ios-apple](https://ews-bridge.webkit.apple.com/builds/sw/T-276/92699e7b-e2c4-41dd-8b79-0fc7c00d5b52/db80031d-80bb-457b-be04-a254c484ef88) 
| | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/130/builds/40567 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/123/builds/48448 "Built successfully") | [❌ 🧪 wpe-wk2](https://ews-build.webkit.org/#/builders/34/builds/91264 "Found 4 new test failures: http/tests/resourceLoadStatistics/website-data-removal-for-site-without-interaction-and-retain-partitioned-and-unpartitioned-storage-for-site-with-interaction.html http/tests/resourceLoadStatistics/website-data-removal-for-site-without-user-interaction-but-with-user-interaction-on-cross-origin-domains.html http/tests/resourceLoadStatistics/website-data-removal-for-site-without-user-interaction.html platform/glib/http/tests/resourceLoadStatistics/website-data-removal-for-site-without-interaction-and-retain-partitioned-and-unpartitioned-storage-for-site-with-interaction.html (failure)") | [❌ 🧪 win-tests](https://ews-build.webkit.org/#/builders/60/builds/60564 "Found 2 new test failures: http/tests/resourceLoadStatistics/website-data-removal-for-site-without-interaction-and-retain-partitioned-and-unpartitioned-storage-for-site-with-interaction.html http/tests/resourceLoadStatistics/website-data-removal-for-site-without-user-interaction-but-with-user-interaction-on-cross-origin-domains.html (failure)") | [❌ 🛠 mac-apple](https://ews-bridge.webkit.apple.com/builds/sw/T-276/92699e7b-e2c4-41dd-8b79-0fc7c00d5b52/6bdef3ac-52c7-41ef-98c7-d90d698d57be) 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/123129 "Passed tests") | [❌ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/132/builds/32401 "Found 1 new test failure: http/tests/resourceLoadStatistics/website-data-removal-for-site-without-interaction-and-retain-partitioned-and-unpartitioned-storage-for-site-with-interaction.html (failure)") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/18/builds/107747 "Passed tests") | [✅ 🧪 api-wpe](https://ews-build.webkit.org/#/builders/41/builds/71816 "Passed tests") | | [❌ 🛠 vision-apple](https://ews-bridge.webkit.apple.com/builds/sw/T-276/92699e7b-e2c4-41dd-8b79-0fc7c00d5b52/d3402efb-8d81-40d3-bd87-ce7e35f86f6d) 
| | [❌ 🧪 ios-wk2-wpt](https://ews-build.webkit.org/#/builders/133/builds/31433 "Exiting early after 10 failures. 49 tests run. 1 failures") | [✅ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/135/builds/25853 "Passed tests") | [✅ 🛠 wpe-cairo](https://ews-build.webkit.org/#/builders/65/builds/70153 "Built successfully") | | 
| | [✅ 🧪 api-ios](https://ews-build.webkit.org/#/builders/13/builds/101881 "Passed tests") | [❌ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/136/builds/26039 "Found 1 new test failure: http/tests/resourceLoadStatistics/website-data-removal-for-site-without-interaction-and-retain-partitioned-and-unpartitioned-storage-for-site-with-interaction.html (failure)") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/129431 "Built successfully") | | 
| | [✅ 🛠 vision](https://ews-build.webkit.org/#/builders/128/builds/47097 "Built successfully") | [❌ 🧪 mac-AS-debug-wk2](https://ews-build.webkit.org/#/builders/122/builds/35731 "Found 1 new test failure: http/tests/resourceLoadStatistics/website-data-removal-for-site-without-interaction-and-retain-partitioned-and-unpartitioned-storage-for-site-with-interaction.html (failure)") | [❌ 🧪 gtk-wk2](https://ews-build.webkit.org/#/builders/1/builds/99881 "Found 4 new test failures: http/tests/resourceLoadStatistics/website-data-removal-for-site-without-interaction-and-retain-partitioned-and-unpartitioned-storage-for-site-with-interaction.html http/tests/resourceLoadStatistics/website-data-removal-for-site-without-user-interaction-but-with-user-interaction-on-cross-origin-domains.html http/tests/resourceLoadStatistics/website-data-removal-for-site-without-user-interaction.html platform/glib/http/tests/resourceLoadStatistics/website-data-removal-for-site-without-interaction-and-retain-partitioned-and-unpartitioned-storage-for-site-with-interaction.html (failure)") | | 
| | [✅ 🛠 vision-sim](https://ews-build.webkit.org/#/builders/121/builds/47463 "Built successfully") | [❌ 🧪 mac-wk2-stress](https://ews-build.webkit.org/#/builders/8/builds/103933 "Exiting early after 10 failures. 42 tests run. 1 new passes 1 failures") | [✅ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/21/builds/99724 "Passed tests") | | 
| | [✅ 🧪 vision-wk2](https://ews-build.webkit.org/#/builders/126/builds/45181 "Passed tests") | [❌ 🧪 mac-intel-wk2](https://ews-build.webkit.org/#/builders/137/builds/23204 "Found 1 new test failure: http/tests/resourceLoadStatistics/website-data-removal-for-site-without-interaction-and-retain-partitioned-and-unpartitioned-storage-for-site-with-interaction.html (failure)") | [  ~~🛠 playstation~~](https://ews-build.webkit.org/#/builders/134/builds/43728 "The change is no longer eligible for processing.") | | 
| | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/127/builds/46959 "Built successfully") | [✅ 🛠 mac-safer-cpp](https://ews-build.webkit.org/#/builders/120/builds/52665 "Built successfully") | | | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/125/builds/46427 "Built successfully") | | | | 
| | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/129/builds/49774 "Built successfully") | | | | 
| | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/124/builds/48111 "Built successfully") | | | | 
<!--EWS-Status-Bubble-End-->

webkit-early-warning-system commented:
Starting EWS tests for https://github.com/WebKit/WebKit/commit/d5857b67890c48afee3b30ddaa78b3031bff90cd. Live statuses available at the PR page, https://github.com/WebKit/WebKit/pull/49807

webkit-early-warning-system commented:
Starting EWS tests for https://github.com/WebKit/WebKit/commit/df7319038b627e462e7b6d771a3f49103e99a5fe. Live statuses available at the PR page, https://github.com/WebKit/WebKit/pull/49807

webkit-early-warning-system commented:
Starting EWS tests for https://github.com/WebKit/WebKit/commit/6a14fe1e9bd68a9f92427dd6953a61c87dd6900a. Live statuses available at the PR page, https://github.com/WebKit/WebKit/pull/49807

webkit-early-warning-system commented:
Starting EWS tests for https://github.com/WebKit/WebKit/commit/92f4adc0a542a735e640c946343e64d568ee84c6. Live statuses available at the PR page, https://github.com/WebKit/WebKit/pull/49807

Commit:
Tracking prevention should only delete script-writable storage associated with a first-party
https://bugs.webkit.org/show_bug.cgi?id=262495
rdar://116144212

Reviewed by NOBODY (OOPS!).

Tracking prevention's current behavior deletes all script-writable storage
without regard for the context where it was created. This means that all
storage associated with an origin will be deleted after 7 days of not visiting
the site as a first-party (in general). This deletion policy is not actually
correct for partitioned third-party storage because that data doesn't have any
connection with visiting the site as a first-party. Instead, we should only
delete script-writable storage associated with the first-party (where the
partition's top-level origin is the first-party).

* LayoutTests/http/tests/resourceLoadStatistics/resources/iframe-with-storage.html: Added.
* LayoutTests/http/tests/resourceLoadStatistics/resources/popup-with-storage-and-iframe.html: Added.
* LayoutTests/http/tests/resourceLoadStatistics/resources/popup-with-storage.html: Added.
* LayoutTests/http/tests/resourceLoadStatistics/resources/util.js:
(sortStringArray):
(addLinebreakToOutput):
(addOutput):
(checkCookies):
(addFrameEventListener):
(request.onupgradeneeded):
(initStorage):
(finish):
(receiveResponse):
(checkFrameStorage):
(checkIDBDataStoreExists):
(checkLocalStorageExists):
(async writeWebsiteDataAndContinue):
(async processWebsiteDataAndContinue):
(async checkWebsiteDataAndContinue):
(async finishTest):
* LayoutTests/http/tests/resourceLoadStatistics/resources/utilities.js: Added.
(setEnableFeature):
(async resetCookiesITP):
(sortStringArray):
(addLinebreakToOutput):
(addOutput):
(finish):
(receiveResponse):
(checkFrameStorage):
(addFrameEventListener):
(onMsg):
(checkPopupCookiesLikeCheckCookies):
(request.onupgradeneeded):
(createIDBDataStore):
(initStorage):
(checkIDBDataStoreExists):
(checkLocalStorageExists):
(once):
(async phaseA_writeWebsiteDataAndReturn.):
(async phaseA_writeWebsiteDataAndReturn):
(async processWebsiteDataAndContinue):
(async checkWebsiteDataAndContinue):
(async finishTest):
* LayoutTests/http/tests/resourceLoadStatistics/website-data-removal-for-site-navigated-to-with-filtered-link-decoration-after-long-deletion.html:
* LayoutTests/http/tests/resourceLoadStatistics/website-data-removal-for-site-with-user-interaction-and-cross-origin-without-user-interaction-expected.txt: Added.
* LayoutTests/http/tests/resourceLoadStatistics/website-data-removal-for-site-with-user-interaction-and-cross-origin-without-user-interaction.html: Added.
* LayoutTests/http/tests/resourceLoadStatistics/website-data-removal-for-site-with-user-interaction-expected.txt:
* LayoutTests/http/tests/resourceLoadStatistics/website-data-removal-for-site-with-user-interaction.html:
* LayoutTests/http/tests/resourceLoadStatistics/website-data-removal-for-site-without-interaction-and-retain-partitioned-and-unpartitioned-storage-for-site-with-interaction-expected.txt: Added.
* LayoutTests/http/tests/resourceLoadStatistics/website-data-removal-for-site-without-interaction-and-retain-partitioned-and-unpartitioned-storage-for-site-with-interaction.html: Added.
* LayoutTests/http/tests/resourceLoadStatistics/website-data-removal-for-site-without-user-interaction-but-with-user-interaction-on-cross-origin-domains-expected.txt: Added.
* LayoutTests/http/tests/resourceLoadStatistics/website-data-removal-for-site-without-user-interaction-but-with-user-interaction-on-cross-origin-domains.html: Added.
* LayoutTests/http/tests/resourceLoadStatistics/website-data-removal-for-site-without-user-interaction-expected.txt:
* LayoutTests/http/tests/resourceLoadStatistics/website-data-removal-for-site-without-user-interaction.html:
* LayoutTests/platform/glib/http/tests/resourceLoadStatistics/website-data-removal-for-site-without-interaction-and-retain-partitioned-and-unpartitioned-storage-for-site-with-interaction.html: Added.
* LayoutTests/platform/glib/http/tests/resourceLoadStatistics/website-data-removal-for-site-without-user-interaction-but-with-user-interaction-on-cross-origin-domains-exppected.txt: Added.
* LayoutTests/platform/glib/http/tests/resourceLoadStatistics/website-data-removal-for-site-without-user-interaction-expected.txt:
* Source/WebKit/NetworkProcess/storage/NetworkStorageManager.cpp:
(WebKit::NetworkStorageManager::deleteDataForRegistrableDomains):
